\documentclass[]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\usepackage{fixltx2e} % provides \textsubscript
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
\else % if luatex or xelatex
  \ifxetex
    \usepackage{mathspec}
  \else
    \usepackage{fontspec}
  \fi
  \defaultfontfeatures{Ligatures=TeX,Scale=MatchLowercase}
\fi
% use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
% use microtype if available
\IfFileExists{microtype.sty}{%
\usepackage{microtype}
\UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\usepackage[margin=1in]{geometry}
\usepackage{hyperref}
\hypersetup{unicode=true,
            pdftitle={Statistical Genetics - 2nd Assignment},
            pdfauthor={Ricard Gardella, Sofia B. Reichl},
            pdfborder={0 0 0},
            breaklinks=true}
\urlstyle{same}  % don't use monospace font for urls
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
\IfFileExists{parskip.sty}{%
\usepackage{parskip}
}{% else
\setlength{\parindent}{0pt}
\setlength{\parskip}{6pt plus 2pt minus 1pt}
}
\setlength{\emergencystretch}{3em}  % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{0}
% Redefines (sub)paragraphs to behave more like sections
\ifx\paragraph\undefined\else
\let\oldparagraph\paragraph
\renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
\let\oldsubparagraph\subparagraph
\renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

%%% Use protect on footnotes to avoid problems with footnotes in titles
\let\rmarkdownfootnote\footnote%
\def\footnote{\protect\rmarkdownfootnote}

%%% Change title format to be more compact
\usepackage{titling}

% Create subtitle command for use in maketitle
\newcommand{\subtitle}[1]{
  \posttitle{
    \begin{center}\large#1\end{center}
    }
}

\setlength{\droptitle}{-2em}

  \title{Statistical Genetics - 2nd Assignment}
    \pretitle{\vspace{\droptitle}\centering\huge}
  \posttitle{\par}
    \author{Ricard Gardella, Sofia B. Reichl}
    \preauthor{\centering\large\emph}
  \postauthor{\par}
      \predate{\centering\large\emph}
  \postdate{\par}
    \date{16th November 2018}


\begin{document}
\maketitle

\subsection{Exercise 1}\label{exercise-1}

\textbf{The file YRIChr1.rda contains genotype information (10000 SNPs)
of individuals from an African population of unrelated individuals. Load
this data into the R environment. The file contains a data objects, X,
with genotype information. This data is in (0,1,2) format, where 0 and 2
represent the homozygotes AA and BB, and 1 represents the heterozygote
AB.}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{library}\NormalTok{(HardyWeinberg)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## Loading required package: mice
\end{verbatim}

\begin{verbatim}
## Loading required package: lattice
\end{verbatim}

\begin{verbatim}
## 
## Attaching package: 'mice'
\end{verbatim}

\begin{verbatim}
## The following objects are masked from 'package:base':
## 
##     cbind, rbind
\end{verbatim}

\begin{verbatim}
## Loading required package: Rsolnp
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{load}\NormalTok{(}\StringTok{"/Users/ricardgardellagarcia/Documents/Master Data science/BSG/Part2/practiques/practica2/YRIChr1.rda"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\subsection{Exercise 2}\label{exercise-2}

\textbf{How many individuals does the database contain?}

We can conclude that we have 107 individuals, same as observations.

\begin{verbatim}
##         rs367896724_AC rs540431307_TA rs555500075_TA
## NA18486              2              0              1
## NA18488              1              0              1
## NA18489              1              0              1
## NA18498              1              0              1
## NA18499              1              0              1
## NA18501              0              0              1
## NA18502              1              0              1
## NA18504              1              0              1
## NA18505              1              0              1
## NA18507              1              0              1
## NA18508              1              0              1
## NA18510              1              0              1
## NA18511              1              0              1
## NA18516              1              0              1
## NA18517              1              0              1
## NA18519              1              0              1
## NA18520              1              0              1
## NA18522              2              0              2
## NA18523              1              0              1
## NA18853              2              0              1
## NA18856              1              0              1
## NA18858              2              0              1
## NA18861              1              0              1
## NA18864              0              0              1
## NA18865              0              0              1
## NA18867              0              0              1
## NA18868              1              0              1
## NA18870              2              0              1
## NA18871              1              0              2
## NA18873              1              0              1
## NA18874              1              0              1
## NA18876              1              0              0
## NA18877              0              0              1
## NA18878              0              0              1
## NA18879              0              0              1
## NA18881              1              0              1
## NA18907              1              0              1
## NA18908              1              0              1
## NA18909              1              0              1
## NA18910              1              0              1
## NA18912              1              0              1
## NA18915              1              0              1
## NA18916              0              0              1
## NA18917              1              0              1
## NA18923              1              0              1
## NA18924              0              0              1
## NA18933              1              0              1
## NA18934              1              0              1
## NA19092              1              0              1
## NA19093              1              0              1
## NA19095              1              0              1
## NA19096              1              0              0
## NA19098              2              0              1
## NA19099              1              0              1
## NA19102              1              0              1
## NA19107              2              0              1
## NA19108              1              0              1
## NA19113              1              0              1
## NA19114              1              0              1
## NA19116              1              0              1
## NA19117              1              0              1
## NA19118              1              0              1
## NA19119              1              0              1
## NA19121              1              0              1
## NA19130              1              0              1
## NA19131              1              0              1
## NA19137              1              0              1
## NA19138              1              0              0
## NA19141              1              0              1
## NA19143              1              0              1
## NA19144              1              0              1
## NA19146              1              0              1
## NA19147              1              0              1
## NA19149              1              0              1
## NA19152              1              0              1
## NA19153              1              0              1
## NA19159              2              0              1
## NA19160              2              0              1
## NA19171              1              0              1
## NA19172              1              0              1
## NA19175              2              0              1
## NA19184              1              0              1
## NA19185              1              0              0
## NA19189              1              0              1
## NA19190              0              0              1
## NA19197              1              0              1
## NA19198              1              0              0
## NA19200              1              0              1
## NA19201              1              0              1
## NA19204              1              0              1
## NA19206              1              0              1
## NA19207              1              0              1
## NA19209              1              0              1
## NA19210              1              0              1
## NA19213              0              0              1
## NA19214              1              0              1
## NA19222              0              0              0
## NA19223              0              0              1
## NA19225              1              0              1
## NA19235              1              0              1
## NA19236              1              0              1
## NA19238              1              0              1
## NA19239              1              0              1
## NA19247              1              0              0
## NA19248              1              0              0
## NA19256              2              0              1
## NA19257              1              0              1
\end{verbatim}

\textbf{What percentage of the variants is monomorphic?}

We have a 69.65\% of monomorphic values as we can see in the results of
the following code.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{data =}\StringTok{ }\DecValTok{0}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\KeywordTok{ncol}\NormalTok{(X))}
\NormalTok{  \{}
    \ControlFlowTok{if}\NormalTok{(}\KeywordTok{sum}\NormalTok{(}\KeywordTok{unique}\NormalTok{(X[i])) }\OperatorTok{!=}\StringTok{ }\DecValTok{0}\NormalTok{) }
\NormalTok{      \{}
\NormalTok{      data =}\StringTok{ }\KeywordTok{cbind}\NormalTok{(data, X[i]) }
\NormalTok{      \}}
\NormalTok{\}}
\NormalTok{data =}\StringTok{ }\NormalTok{data[}\OperatorTok{-}\DecValTok{1}\NormalTok{]}
\NormalTok{(}\KeywordTok{ncol}\NormalTok{(X) }\OperatorTok{-}\StringTok{ }\KeywordTok{ncol}\NormalTok{(data)) }\OperatorTok{/}\DecValTok{100} 
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 69.65
\end{verbatim}

\textbf{How many variants remain in the database?}

The new data set have 3035 variants remaining.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{ncol}\NormalTok{(data)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 3035
\end{verbatim}

\textbf{Determine the genotype counts for these variants, and store them
in matrix.}

In order to determine the genotypes, we create a matrix of 3 rows
(AA,AB,BB). Then, we count all the apperances of an AA,AB and BB of
every SNP. It is not necessary to label the rows as the library takes as
default values that the rows will be AA, AB and BB, in this order.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{matrixvariants =}\StringTok{ }\KeywordTok{matrix}\NormalTok{(}\DataTypeTok{data =} \DecValTok{0}\NormalTok{,}\DataTypeTok{nrow =} \DecValTok{3}\NormalTok{,}\DataTypeTok{ncol =} \KeywordTok{ncol}\NormalTok{(data))}
\ControlFlowTok{for}\NormalTok{ (i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\KeywordTok{ncol}\NormalTok{(data))}
\NormalTok{\{}
\NormalTok{  matrixvariants[}\DecValTok{1}\NormalTok{,i] =}\StringTok{ }\KeywordTok{sum}\NormalTok{(data[i] }\OperatorTok{==}\StringTok{ }\DecValTok{0}\NormalTok{)}
\NormalTok{  matrixvariants[}\DecValTok{2}\NormalTok{,i] =}\StringTok{ }\KeywordTok{sum}\NormalTok{(data[i] }\OperatorTok{==}\StringTok{ }\DecValTok{1}\NormalTok{)}
\NormalTok{  matrixvariants[}\DecValTok{3}\NormalTok{,i] =}\StringTok{ }\KeywordTok{sum}\NormalTok{(data[i] }\OperatorTok{==}\StringTok{ }\DecValTok{2}\NormalTok{)}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\textbf{Apply a chi-square test without continuity correction for
Hardy-Weinberg equilibrium to each SNP. How many SNPs are significant
(use a = 0:05)?}

We can observe that 1511 are in, the other ones are rejected.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{table}\NormalTok{(HWChisq }\OperatorTok{>=}\StringTok{ }\FloatTok{0.05}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## FALSE  TRUE 
##  1524  1511
\end{verbatim}

\subsection{Exercise 3}\label{exercise-3}

\textbf{How many markers of the remaining non-monomorphic markers would
you expect to be out of equilibrium by the elect of chance alone?}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{round}\NormalTok{(}\KeywordTok{nrow}\NormalTok{(tmatrixvariants) }\OperatorTok{*}\StringTok{ }\FloatTok{0.05}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 152
\end{verbatim}

\subsection{Exercise 4}\label{exercise-4}

\textbf{Apply an Exact test for Hardy-Weinberg equilibrium to each SNP.
You can use function HWExactStats for fast computation. How many SNPs
are signifcant (use a = 0.05). Is the result consistent with the
chi-square test?}

Now 2909 pass the test, quite different than before, the results are not
consistent with the chi-square test as we can observe in this results.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{HWExact <-}\StringTok{ }\KeywordTok{HWExactStats}\NormalTok{(tmatrixvariants)}
\KeywordTok{table}\NormalTok{(HWExact }\OperatorTok{>=}\StringTok{ }\FloatTok{0.05}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 
## FALSE  TRUE 
##   126  2909
\end{verbatim}

\subsection{Exercise 5}\label{exercise-5}

\textbf{Apply a likelihood ratio test for Hardy-Weinberg equilibrium to
each SNP, using the HWLratio function. How many SNPs are significant
(use a = 0:05). Is the result consistent with the chi-square test?}

We can see that in this case we do not have an Stats function, so we
crafted our own function. In this case the results are similar with the
Exact test but not with the Chi-Square test, as we have 2890 p-values
that pass the test.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{pvaluesratio}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 2890
\end{verbatim}

\subsection{Exercise 6}\label{exercise-6}

\textbf{Apply a permutation test for Hardy-Weinberg equilibrium to the
first 10 SNPs, using the classical chi-square test (without continuity
correction) as a test statistic. List the 10 p-values, together with the
10 p-values of the exact tests. Are the result consistent?}

\begin{verbatim}
##       [,1] [,2] [,3]
##  [1,]   13   83   11
##  [2,]    8   97    2
##  [3,]  104    3    0
##  [4,]  105    2    0
##  [5,]   80   25    2
##  [6,]   80   25    2
##  [7,]  103    4    0
##  [8,]  106    1    0
##  [9,]  104    3    0
## [10,]  104    3    0
\end{verbatim}

\begin{verbatim}
##       [,1] [,2] [,3]  [,4]
##  [1,]   13   83   11 0.000
##  [2,]    8   97    2 0.000
##  [3,]  104    3    0 0.883
##  [4,]  105    2    0 0.922
##  [5,]   80   25    2 0.977
##  [6,]   80   25    2 0.977
##  [7,]  103    4    0 0.844
##  [8,]  106    1    0 0.961
##  [9,]  104    3    0 0.883
## [10,]  104    3    0 0.883
\end{verbatim}

\begin{verbatim}
##       [,1] [,2] [,3]  [,4] [,5]
##  [1,]   13   83   11 0.000    0
##  [2,]    8   97    2 0.000    0
##  [3,]  104    3    0 0.883    1
##  [4,]  105    2    0 0.922    1
##  [5,]   80   25    2 0.977    1
##  [6,]   80   25    2 0.977    1
##  [7,]  103    4    0 0.844    1
##  [8,]  106    1    0 0.961    1
##  [9,]  104    3    0 0.883    1
## [10,]  104    3    0 0.883    1
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{newmatrix}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##       [,1] [,2] [,3]  [,4] [,5]
##  [1,]   13   83   11 0.000    0
##  [2,]    8   97    2 0.000    0
##  [3,]  104    3    0 0.883    1
##  [4,]  105    2    0 0.922    1
##  [5,]   80   25    2 0.977    1
##  [6,]   80   25    2 0.977    1
##  [7,]  103    4    0 0.844    1
##  [8,]  106    1    0 0.961    1
##  [9,]  104    3    0 0.883    1
## [10,]  104    3    0 0.883    1
\end{verbatim}

\subsection{Exercise 7}\label{exercise-7}

\textbf{Depict all SNPs simultaeneously in a ternary plot, and comment
on your result (because many genotype counts repeat, you may use
UniqueGenotypeCounts to speed up the computations)}

In the plot we can see that half of the ternary plot is empty. We used
unique genotype as requested.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{HWTernaryPlot}\NormalTok{(}\KeywordTok{UniqueGenotypeCounts}\NormalTok{(tmatrixvariants)[}\DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 3035 rows in X
## 456 unique rows in X
\end{verbatim}

\includegraphics{BSG_2HWE_Gardella_Reichl_files/figure-latex/plot-1.pdf}

\subsection{Exercise 8}\label{exercise-8}

\textbf{Can you explain why half of the ternary diagram is empty?}

Because A appear aproximately three times more than B. Also we can
observe that there is no case where B appears more than A. So that's why
we see half of the diagram empty.

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{unique <-}\StringTok{ }\KeywordTok{UniqueGenotypeCounts}\NormalTok{(tmatrixvariants)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## 3035 rows in X
## 456 unique rows in X
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{head}\NormalTok{(unique[}\DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{]); }\KeywordTok{tail}\NormalTok{(unique[}\DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{])}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##    AA AB BB
## 1 106  1  0
## 2 105  2  0
## 3 104  3  0
## 4 103  4  0
## 5 102  5  0
## 6 101  6  0
\end{verbatim}

\begin{verbatim}
##     AA AB BB
## 451 28 52 27
## 452 35 44 28
## 453 31 48 28
## 454 31 46 30
## 455 31 45 31
## 456 37 35 35
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(A <-}\StringTok{ }\KeywordTok{sum}\NormalTok{(unique[,}\DecValTok{1}\NormalTok{])}\OperatorTok{+}\DecValTok{1}\OperatorTok{/}\DecValTok{2}\OperatorTok{*}\KeywordTok{sum}\NormalTok{(unique[,}\DecValTok{2}\NormalTok{]))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 35825
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{(B <-}\StringTok{ }\KeywordTok{sum}\NormalTok{(unique[,}\DecValTok{3}\NormalTok{])}\OperatorTok{+}\DecValTok{1}\OperatorTok{/}\DecValTok{2}\OperatorTok{*}\KeywordTok{sum}\NormalTok{(unique[,}\DecValTok{2}\NormalTok{]))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] 12967
\end{verbatim}

\section{Exercise 9}\label{exercise-9}

\textbf{Make a histogram of the p-values obtained in the chi-square
test. What distribution would you expect if HWE would hold for the data
set?}\\
I would say that it is similar to a Beta distribution with high
\(\alpha\) and low value of \(\beta\) parameters.**\\
\includegraphics{BSG_2HWE_Gardella_Reichl_files/figure-latex/unnamed-chunk-9-1.pdf}

\begin{verbatim}
##        0%       25%       50%       75%      100% 
## 0.4048184 0.8708210 0.9319513 0.9715292 0.9999819
\end{verbatim}

\textbf{Make a Q-Q plot of the p values obtained in the chi-square test
against the quantiles of the distribution that you consider relevant.
What is your conclusion?}

\includegraphics{BSG_2HWE_Gardella_Reichl_files/figure-latex/unnamed-chunk-11-1.pdf}

Well, it seems that this distribution is not that similar as we thought.

\section{Exercise 10}\label{exercise-10}

\textbf{Imagine that for a particular marker the counts of the two
homozygotes are accidentally interchanged. Would this affect the
statistical tests for HWE? Try it on the computer if you want. Argue
your answer}

No,because the equilibrium will be kept, as we are only changing the
``labels'' of the homozygotes.

\section{Exercise 11}\label{exercise-11}

\textbf{Compute the inbreeding coeffcient (\(f\)) for each SNP, and make
a histogram of \(f\). You can use function HWf for this purpose. Give
descriptive statistics (mean, standard deviation, etc) of \(f\)
calculated over the set of SNPs.}

\textbf{What distribution do you expect \(f\) to follow
theoretically?}\\
In this case it seems that the inbreeding coefficient follows something
like a Normal Distribution. To adjust it we will take as the mean the
same as the inbreeding coefficient distribution. As we have seen the
standard desviation seem too large so we will use another one: 0.05 as
we know that most of the values are overall the mean.

\includegraphics{BSG_2HWE_Gardella_Reichl_files/figure-latex/unnamed-chunk-12-1.pdf}

\textbf{Use a probability plot to confirm your idea.}\\
\includegraphics{BSG_2HWE_Gardella_Reichl_files/figure-latex/unnamed-chunk-13-1.pdf}

\section{Exercise 12}\label{exercise-12}

\textbf{Make a plot of the observed chi-square statistics against the
inbreeding coeffcient (\(f\)). What do you observe?}\\
\includegraphics{BSG_2HWE_Gardella_Reichl_files/figure-latex/unnamed-chunk-14-1.pdf}

It seems that when the embredding coefficient is 0 then the chi-squared
statistic is also 0. When the value of the imbreeding coefficient
increases positive and negatively then the chi-squared statistic also
increases. So, it seem they are quite correlated.

\textbf{Can you give an equation that relates the two statistics?}\\
As we have plotted before, one possible equation could be \(x=y^2*100\)
--\textgreater{} \(y=\sqrt{x/100}\).

\section{Exercise 13}\label{exercise-13}

\textbf{Simulate SNPs under the assumption of Hardy-Weinberg
equilibrium. Simulate the SNPs of this database, and take care to match
each of the SNPs in your database with a simulated SNP that has the same
sample size and allele frequency. You can use function HWData of the
HardyWeinberg package for this purpose.}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{freq <-}\StringTok{ }\KeywordTok{numeric}\NormalTok{(}\KeywordTok{nrow}\NormalTok{(tmatrixvariants))}
\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\KeywordTok{nrow}\NormalTok{(tmatrixvariants))\{}
\NormalTok{  aa <-tmatrixvariants[i,}\DecValTok{1}\NormalTok{]}
\NormalTok{  ab <-tmatrixvariants[i,}\DecValTok{2}\NormalTok{]}
\NormalTok{  bb <-tmatrixvariants[i,}\DecValTok{3}\NormalTok{]}
\NormalTok{  freq[i] =}\StringTok{ }\NormalTok{(aa}\OperatorTok{+}\NormalTok{ab}\OperatorTok{/}\DecValTok{2}\NormalTok{)}\OperatorTok{/}\NormalTok{(aa}\OperatorTok{+}\NormalTok{ab}\OperatorTok{+}\NormalTok{bb)}
\NormalTok{\}}

\NormalTok{simul <-}\StringTok{ }\KeywordTok{data.frame}\NormalTok{(}\KeywordTok{HWData}\NormalTok{(}\DataTypeTok{nm=}\KeywordTok{nrow}\NormalTok{(tmatrixvariants),}\DataTypeTok{n=}\DecValTok{107}\NormalTok{,}\DataTypeTok{p=}\NormalTok{freq))}
\ControlFlowTok{for}\NormalTok{(i }\ControlFlowTok{in} \DecValTok{1}\OperatorTok{:}\KeywordTok{nrow}\NormalTok{(simul))\{}
\NormalTok{  stats <-}\StringTok{ }\KeywordTok{HWChisq}\NormalTok{(}\KeywordTok{as.numeric}\NormalTok{(simul[i, }\DecValTok{1}\OperatorTok{:}\DecValTok{3}\NormalTok{]),}\DataTypeTok{cc=}\DecValTok{0}\NormalTok{, }\DataTypeTok{verbose =} \OtherTok{FALSE}\NormalTok{)}
\NormalTok{  simul}\OperatorTok{$}\NormalTok{chisq[i] <-}\StringTok{ }\NormalTok{stats[}\DecValTok{1}\NormalTok{]}
\NormalTok{  simul}\OperatorTok{$}\NormalTok{pval[i] <-}\StringTok{ }\NormalTok{stats[}\DecValTok{2}\NormalTok{]}
\NormalTok{  simul}\OperatorTok{$}\NormalTok{D[i] <-}\StringTok{ }\NormalTok{stats[}\DecValTok{3}\NormalTok{]}
\NormalTok{  simul}\OperatorTok{$}\NormalTok{p[i] <-}\StringTok{ }\NormalTok{stats[}\DecValTok{4}\NormalTok{]}
\NormalTok{  simul}\OperatorTok{$}\NormalTok{f[i] <-}\StringTok{ }\NormalTok{stats[}\DecValTok{5}\NormalTok{]}
\NormalTok{  simul}\OperatorTok{$}\NormalTok{expected[i] <-}\StringTok{ }\NormalTok{stats[}\DecValTok{6}\NormalTok{]}
\NormalTok{\}}
\end{Highlighting}
\end{Shaded}

\textbf{Compare the distribution of the observed chi-square statistics
with the distribution of the chi-square statistics of the simulated SNPs
by making a Q-Q plot. What do you observe? State your conclusions.}\\
\includegraphics{BSG_2HWE_Gardella_Reichl_files/figure-latex/unnamed-chunk-16-1.pdf}
It seems that there are some of the variants that are different in the
simulated dataset than in the original one.

\section{Exercise 14}\label{exercise-14}

\textbf{We reconsider the exact test for HWE, using different
signifocant levels. Report the number and percentage of significant
variants using an exac test for HWE with \(\alpha\) = 0.10; 0.05; 0.01
and 0.001. State your conclusions.}\\
We will calculate the Exact test for the variants matrix and we will see
how the results are:

\begin{verbatim}
## Table with alpha=0.1:
\end{verbatim}

\begin{verbatim}
## 
## Greater than 0.1   Lower than 0.1 
##             2852              183
\end{verbatim}

\begin{verbatim}
## 
## % Greater than 0.1   % Lower than 0.1 
##          93.970346           6.029654
\end{verbatim}

\begin{verbatim}
## -------------------------------------------------------------------------
\end{verbatim}

\begin{verbatim}
## Table with alpha=0.05:
\end{verbatim}

\begin{verbatim}
## 
## Greater than 0.05   Lower than 0.05 
##              2909               126
\end{verbatim}

\begin{verbatim}
## 
## % Greater than 0.05   % Lower than 0.05 
##           95.848435            4.151565
\end{verbatim}

\begin{verbatim}
## -------------------------------------------------------------------------
\end{verbatim}

\begin{verbatim}
## Table with alpha=0.01:
\end{verbatim}

\begin{verbatim}
## 
## Greater than 0.01   Lower than 0.01 
##              2952                83
\end{verbatim}

\begin{verbatim}
## 
## % Greater than 0.01   % Lower than 0.01 
##           97.265239            2.734761
\end{verbatim}

\begin{verbatim}
## -------------------------------------------------------------------------
\end{verbatim}

\begin{verbatim}
## Table with alpha=0.001:
\end{verbatim}

\begin{verbatim}
## 
## Greater than 0.001   Lower than 0.001 
##               2986                 49
\end{verbatim}

\begin{verbatim}
## 
## % Greater than 0.001   % Lower than 0.001 
##            98.385502             1.614498
\end{verbatim}

\section{Exercise 15}\label{exercise-15}

\textbf{Do you think genotyping error is a problem for the database you
just studied? Explain your opinion.}\\
By taking a look at the exact test we can see that there are 126
variants that are out of eqilibrium (with a normal alpha of 0.05). If we
compare the variants that are out of equilibrium from the ChiSquared
test and the Exact test the difference is quite notorious.

\begin{verbatim}
## Result of HWChisq pvalues of the variants:
\end{verbatim}

\begin{verbatim}
## 
## Greater than 0.05   Lower than 0.05 
##              1511              1524
\end{verbatim}

\begin{verbatim}
## -------------------------------------------------------------------------
\end{verbatim}

\begin{verbatim}
## Result of HWExact pvalues of the variants:
\end{verbatim}

\begin{verbatim}
## 
## Greater than 0.05   Lower than 0.05 
##              2909               126
\end{verbatim}

So, we can assume that this is due to genotyping error.


\end{document}
