---
title: "Statistical Genetics - 2nd Assignment"
author: "Ricard Gardella, Sofia B. Reichl"
date: "16th November 2018"
output:
  html_document:
    df_print: paged
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Exercise 1
**The file YRIChr1.rda contains genotype information (10000 SNPs) of individuals from an African
population of unrelated individuals. Load this data into the R environment. The file contains a
data objects, X, with genotype information. This data is in (0,1,2) format, where 0 and 2 represent the homozygotes AA and BB, and 1 represents the heterozygote AB.**

```{r load}
library(HardyWeinberg)
load("/Users/ricardgardellagarcia/Documents/Master Data science/BSG/Part2/practiques/practica2/YRIChr1.rda")

```

## Exercise 2

**How many individuals does the database contain?**

We can conclude that we have 107 individuals, same as observations.
```{r pressure, echo=FALSE}
X[1:3]
```

**What percentage of the variants is monomorphic?**

We have a 69.65% of monomorphic values as we can see in the results of the following code.

```{r perofmonomorphic}
data = 0
for (i in 1:ncol(X))
  {
    if(sum(unique(X[i])) != 0) 
      {
      data = cbind(data, X[i]) 
      }
}
data = data[-1]
(ncol(X) - ncol(data)) /100 
```

**How many variants remain in the database?**

The new data set have 3035 variants remaining.
```{r}
ncol(data)
```

**Determine the genotype counts for these variants, and store them in matrix.**

In order to determine the genotypes, we create a matrix of 3 rows (AA,AB,BB). Then, we count all the apperances of an AA,AB and BB of every SNP. It is not necessary to label the rows as the library takes as default values that the rows will be AA, AB and BB, in this order.
```{r}
matrixvariants = matrix(data = 0,nrow = 3,ncol = ncol(data))
for (i in 1:ncol(data))
{
  matrixvariants[1,i] = sum(data[i] == 0)
  matrixvariants[2,i] = sum(data[i] == 1)
  matrixvariants[3,i] = sum(data[i] == 2)
}
```

**Apply a chi-square test without continuity correction for Hardy-Weinberg equilibrium to each SNP. How many SNPs are significant (use a = 0:05)?**

We can observe that 1511 are in, the other ones are rejected.

```{r test1, echo=FALSE}
tmatrixvariants <- t(matrixvariants)
HWChisq <- HWChisqStats(tmatrixvariants)
```

```{r}
table(HWChisq >= 0.05)
```

## Exercise 3

**How many markers of the remaining non-monomorphic markers would you expect to be out of equilibrium by the elect of chance alone?**

```{r}
round(nrow(tmatrixvariants) * 0.05)
```


## Exercise 4

**Apply an Exact test for Hardy-Weinberg equilibrium to each SNP. You can use function HWExactStats for fast computation. How many SNPs are signifcant (use a = 0.05). Is the result
consistent with the chi-square test?**

Now 2909 pass the test, quite different than before, the results are not consistent with the chi-square test as we can observe in this results.
```{r}
HWExact <- HWExactStats(tmatrixvariants)
table(HWExact >= 0.05)
```

##Exercise 5

**Apply a likelihood ratio test for Hardy-Weinberg equilibrium to each SNP, using the HWLratio
function. How many SNPs are significant (use a = 0:05). Is the result consistent with the chi-square
test?**

We can see that in this case we do not have an Stats function, so we crafted our own function. In this case the results are similar with the Exact test but not with the Chi-Square test, as we have 2890 p-values that pass the test.

```{r ex5, echo=FALSE,warning=FALSE,include=FALSE}
pvaluesratio = 0
HWratio <- function(x)
{
  test = HWLratio(x)
  if(test$pval >= 0.05)
  {
    pvaluesratio <<- pvaluesratio + 1
  }
}
apply(matrixvariants, 2, HWratio)
pvaluesratio
```

```{r}
pvaluesratio
```

##Exercise 6

**Apply a permutation test for Hardy-Weinberg equilibrium to the first 10 SNPs, using the
classical chi-square test (without continuity correction) as a test statistic. List the 10 p-values, together with the 10 p-values of the exact tests. Are the result consistent?**

```{r ex6, echo=FALSE}
(newmatrix <- tmatrixvariants[1:10,])
(newmatrix <- cbind(newmatrix, round(HWChisqStats(newmatrix,pvalues=T),3)))
(newmatrix <- cbind(newmatrix, round(HWExactStats(newmatrix[,1:3]),3)))
```

```{r}
newmatrix
```

##Exercise 7

**Depict all SNPs simultaeneously in a ternary plot, and comment on your result (because many
genotype counts repeat, you may use UniqueGenotypeCounts to speed up the computations)**


In the plot we can see that half of the ternary plot is empty. We used unique genotype as requested. 

```{r plot}
HWTernaryPlot(UniqueGenotypeCounts(tmatrixvariants)[1:3])
```

## Exercise 8

**Can you explain why half of the ternary diagram is empty?**

Because A appear aproximately three times more than B. Also we can observe that there is no case where B appears more than A. So that's why we see half of the diagram empty.

```{r}
unique <- UniqueGenotypeCounts(tmatrixvariants)
head(unique[1:3]); tail(unique[1:3])
(A <- sum(unique[,1])+1/2*sum(unique[,2]))
(B <- sum(unique[,3])+1/2*sum(unique[,2]))
```



# Exercise 9
**Make a histogram of the p-values obtained in the chi-square test. What distribution would you expect if HWE would hold for the data set?**  
I would say that it is similar to a Beta distribution with high $\alpha$ and low value of $\beta$ parameters.**  
```{r echo=F}
betadist <- rbeta(nrow(tmatrixvariants),10,1)
par(mfrow=c(1,2))
hist(HWChisqStats(tmatrixvariants,pvalues=T))
hist(betadist)
```

```{r echo=F}
quantile(betadist)
```

**Make a Q-Q plot of the p values obtained in the chi-square test against the quantiles of the distribution that you consider relevant. What is your conclusion?** 

```{r echo=F}
qqnorm(HWChisqStats(tmatrixvariants,pvalues=T))
qqline(quantile(betadist))
```

Well, it seems that this distribution is not that similar as we thought.


# Exercise 10
**Imagine that for a particular marker the counts of the two homozygotes are accidentally interchanged. Would this affect the statistical tests for HWE? Try it on the computer if you want. Argue your answer**  

No,because the equilibrium will be kept, as we are only changing the "labels" of the homozygotes.



# Exercise 11
**Compute the inbreeding coeffcient ($f$) for each SNP, and make a histogram of $f$. You can use function HWf for this purpose. Give descriptive statistics (mean, standard deviation, etc) of $f$ calculated over the set of SNPs.**  

```{r ex11, include=FALSE, warning=FALSE}
inbreed <- apply(tmatrixvariants,1,HWf)
summary(inbreed);
sd(inbreed)
```

**What distribution do you expect $f$ to follow theoretically?**  
In this case it seems that the inbreeding coefficient follows something like a Normal Distribution. To adjust it we will take as the mean the same as the inbreeding coefficient distribution. As we have seen the standard desviation seem too large so we will use another one: 0.05 as we know that most of the values are overall the mean.

```{r echo=F}
par(mfrow=c(1,2))
hist(inbreed,breaks=100,main='Inbreeding Coeff.',xlab='Inbreeding coefficient')
hist(rnorm(length(inbreed),mean(inbreed),0.05),xlim=c(-1,1),breaks=100,xlab='Inbreeding Approximation',main='Inbreeding Coeff. Approximation')
```

**Use a probability plot to confirm your idea.**  
```{r echo=F}
par(mfrow=c(1,1))
qqplot(inbreed,rnorm(length(inbreed),0,0.05),main='QQ plot of real vs. approximated',xlab='Real Inbreeding Coeff.',ylab='Approximated Inbreeding Coeff.')
```


# Exercise 12
**Make a plot of the observed chi-square statistics against the inbreeding coeffcient ($f$). What do you observe?**  
```{r echo=F, warning=F}
chisq <- c()
for(i in 1:nrow(tmatrixvariants)){
  chisq[i]<-as.numeric(HWChisq(tmatrixvariants[i,],verbose=F)[1])
}
par(mfrow=c(1,2))
plot(chisq,inbreed,xlab='Chi-Square Statistics',ylab='Inbreeding Coefficient',main='Chi-Square vs. inbreeding coeff.')
x <- seq(-1,1,0.01)
plot(x^2*100,x)
```

It seems that when the embredding coefficient is 0 then the chi-squared statistic is also 0. When the value of the imbreeding coefficient increases positive and negatively then the chi-squared statistic also increases. So, it seem they are quite correlated.

**Can you give an equation that relates the two statistics?**  
As we have plotted before, one possible equation could be $x=y^2*100$ --> $y=\sqrt{x/100}$.


# Exercise 13
**Simulate SNPs under the assumption of Hardy-Weinberg equilibrium. Simulate the SNPs of this database, and take care to match each of the SNPs in your database with a simulated SNP that has the same sample size and allele frequency. You can use function HWData of the HardyWeinberg package for this purpose.**  
```{r warning=F}
freq <- numeric(nrow(tmatrixvariants))
for(i in 1:nrow(tmatrixvariants)){
  aa <-tmatrixvariants[i,1]
  ab <-tmatrixvariants[i,2]
  bb <-tmatrixvariants[i,3]
  freq[i] = (aa+ab/2)/(aa+ab+bb)
}

simul <- data.frame(HWData(nm=nrow(tmatrixvariants),n=107,p=freq))
for(i in 1:nrow(simul)){
  stats <- HWChisq(as.numeric(simul[i, 1:3]),cc=0, verbose = FALSE)
  simul$chisq[i] <- stats[1]
  simul$pval[i] <- stats[2]
  simul$D[i] <- stats[3]
  simul$p[i] <- stats[4]
  simul$f[i] <- stats[5]
  simul$expected[i] <- stats[6]
}
```

**Compare the distribution of the observed chi-square statistics with the distribution of the chi-square statistics of the simulated SNPs by making a Q-Q plot. What do you observe? State your conclusions.**  
```{r echo=F}
qqplot(unlist(HWExactStats(tmatrixvariants)), unlist(simul$pval), xlab="Original Data", ylab="Simulated Data")
```
It seems that there are some of the variants that are different in the simulated dataset than in the original one.


# Exercise 14
**We reconsider the exact test for HWE, using different signifocant levels. Report the number and percentage of significant variants using an exac test for HWE with $\alpha$ = 0.10; 0.05; 0.01 and 0.001. State your conclusions.**  
We will calculate the Exact test for the variants matrix and we will see how the results are:
```{r echo=F}
results=HWExactStats(tmatrixvariants)
  
cat('Table with alpha=0.1:');
  table(ifelse(results<0.1,'Lower than 0.1','Greater than 0.1'))
  table(ifelse(results<0.1,'% Lower than 0.1','% Greater than 0.1'))/nrow(tmatrixvariants)*100
cat('-------------------------------------------------------------------------')
cat('Table with alpha=0.05:');
  table(ifelse(results<0.05,'Lower than 0.05','Greater than 0.05'))
  table(ifelse(results<0.05,'% Lower than 0.05','% Greater than 0.05'))/nrow(tmatrixvariants)*100
cat('-------------------------------------------------------------------------')
cat('Table with alpha=0.01:');
  table(ifelse(results<0.01,'Lower than 0.01','Greater than 0.01'))
  table(ifelse(results<0.01,'% Lower than 0.01','% Greater than 0.01'))/nrow(tmatrixvariants)*100
cat('-------------------------------------------------------------------------')
cat('Table with alpha=0.001:');
  table(ifelse(results<0.001,'Lower than 0.001','Greater than 0.001'))
  table(ifelse(results<0.001,'% Lower than 0.001','% Greater than 0.001'))/nrow(tmatrixvariants)*100

```


# Exercise 15
**Do you think genotyping error is a problem for the database you just studied? Explain your opinion.**  
By taking a look at the exact test we can see that there are 126 variants that are out of eqilibrium (with a normal alpha of 0.05). If we compare the variants that are out of equilibrium from the ChiSquared test and the Exact test the difference is quite notorious.  
```{r echo=F}
cat('Result of HWChisq pvalues of the variants:');table(ifelse(HWChisqStats(tmatrixvariants)<0.05,'Lower than 0.05','Greater than 0.05'))
cat('-------------------------------------------------------------------------')
cat('Result of HWExact pvalues of the variants:');table(ifelse(HWExactStats(tmatrixvariants)<0.05,'Lower than 0.05','Greater than 0.05'))
```
So, we can assume that this is due to genotyping error.
