---
title: "BSG - Haplotype estimation"
author: "Ricard Gardella, Sofia B. Reichl"
date: "5th December 2018"
output:
  html_document:
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(HardyWeinberg)
library(genetics)
library(ggplot2)
library(gridExtra)
library(LDheatmap)
library(haplo.stats)
set.seed(123)
```

# Exercise 1

**The file SNPChr20.rda contains genotype information of 310 individuals of unknown background.
The genotype information concerns 50.000 SNPs on chromosome 20. Load this data into the R
environment. The data file contains a matrix Y containing the allele counts (0,1 or 2) for 50.000
SNPs for one of the alleles of each SNP.**

```{r load}
load('/Users/ricardgardellagarcia/Documents/Master Data science/BSG/Part2/practiques/practica5/SNPChr20.rda')
```

# Exercise 2

**Compute the Manhattan distance matrix between the 310 individuals (this may take a few
minutes) Include a submatrix of dimension 5 by 5 with the distances between the first 5 individuals in your report**

```{r}
Y.manhattan = dist(Y,method = "manhattan",diag = FALSE)
Y.manhattan = as.matrix(Y.manhattan)
print("The first 5 rows of the Manhattan distance are:")
Y.manhattan[1:5,1:5]
```

# Execise 3 

**Use metric multidimensional scaling to obtain a map of the individuals, and include your map
in your report. Do you think the data come from one homogeneous population?**

We can see in a very clear way that the data is not homogeneous. This data come for three different populations as we can see in the plot.

```{r}
Y.dist = dist(Y)
Y.dist = as.matrix(Y.dist)
mds.out <- cmdscale(Y.dist,k=nrow(Y)-1,eig=TRUE)
X <- mds.out$points[,1:2]
plot(X[,2],X[,1], xlab="", ylab="", main="Exercise 3",asp=1)
```

#Exercise 4
**Report the first 10 eigenvalues of the solution.**

Here are the first 10 eigenvalues of the solution.
```{r}
mds.out$eig[1:10]
```

#Exercise 5

** Does a perfect representation of this n x n distance matrix exist, in n or fewer dimensions? **

No idea...
```{r}

```

#Exercise 6 
 **What is the goodness-of-fit of a two-dimensional approximation to your distance matrix? **
 
 Here we report the GoF using k=2 in the distance matrix.
```{r}
gof <- cmdscale(Y.dist, k=2, eig=TRUE)
gof <- gof$GOF
print(round(gof,digits=3))

```
#Exercise 7

**Make a plot of the estimated distances (according to your map of individuals) versus the
observed distances. Regress estimated distances on observed distances and report the coefficient of determination of the regression.**

Here we see the differnece between the observed and the fitted values in the plot.

```{r}
X.dist = as.matrix(dist(X))

x.1 <- Y.dist[lower.tri(Y.dist)] 
x.2 <- X.dist[lower.tri(X.dist)] 
plot(x.1, x.2, xlab="Observed", ylab="fitted")

```

